### PR TITLE
add clap like env variable hint

### PR DESCRIPTION
**Pull request by glihm**
A suggestion to have clap like hint for the `STARKNET_KEYSTORE` env variable.
Related to #4.

### CODE REVIEW

**Comment by xJonathanLEI** [comment]
Thanks for working on this. However, I'm not sure if this is the correct way to go. Normally this "env var hint" would automatically show the value if already set, but this one obviously doesn't.

**Comment by glihm** [comment]
Yes exactly, based on your comment inside the code I have tried this solution as we can't have the hint from clap.

From a user experience, it's useful to have the hint, this was the motivation of my commit.

But I totally understand your take on this, what do you suggest then? No hint for now or do we have to investigate how to handle this with clap for this particular case? (It would be the better if it's possible)

**Comment by xJonathanLEI** [comment]
I don't think it's possible. Pretty sure `clap` excepts a static `str` here but I could be wrong obviously.

My concern is that this breaks user assumption. For anyone who's aware of the "value shown in hints when set" behavior, this creates an unwanted exception where users would be left asking themselves why the value is not detected.

I think a hint that says this can be set via the env var in plain English without attempting to mimic the `clap` format is better.

**Comment by xJonathanLEI** [comment]
To avoid confusion, let me clarify a bit on what I meant.

So when the value exists clap is able to show something like (if using `env` derive):

```
[STARKNET_KEYSTORE=/path/to/key]
```

but with our fake hint it's always just:

```
[STARKNET_KEYSTORE=]
```

which breaks user assumption on how this feature works.

**Comment by xJonathanLEI** [comment]
I think I'm still not clear. So the desirable behavior here is for the CLI to reject these:

- `--keystore xxx --private-key xxx`
- `--keystore-password xxx --private-key xxx`

and to be clear, this is _not_ a precedence problem. I believe option conflicts (like the above) should be explicitly rejected instead of having precedence applied quietly (read: magically).

But this should be allowed:

- `STARKNET_KEYSTORE=xxx ... --private-key xxx`

My rationale is that, with environment variables we're encouraging users to set global defaults with their shell profiles, which is nice since each invocation becomes much simpler, and you _can_ indeed override the keystore with another keystore file using the flag version of `--keystore`. However, imagine say now you want to send a test transaction with a raw private key for development purpose - oops, a signer option conflict.

This means we need to distinguish between whether the `keystore` option itself comes from environment variable or command line. For env var input we interpret it as a global default that can be safely ignored, and for the latter we would interpret it as a user mistake.

I hope this is clear. With the latest diffs in this PR I think you're just reverting this whole "env var does not conflict but flag does" functionality, which means that users wouldn't be able to use `--private-key` when a global `STARKNET_KEYSTORE` is set - not ideal.

**Comment by glihm** [comment]
Thanks a lot for this explanation and your vision @xJonathanLEI.
I've then pushed a last attempt that may fit your vision.

The only point of this solution is that the following scenario will not clash:
```bash
export STARKNET_KEYSTORE=~/path1/key.json
starkli declare --keystore ~/path1/key.json --private-key 0x1234 contract.json
```

But otherwise, it will clash as expected.

As you mentioned, the way we use clap with the static strings is limiting us from using clap methods to know if the argument was provided or loaded from the env.

**Comment by glihm** [comment]
> Actually, I think I'm wrong on the static str requirement. There is a `string` feature in `clap` that allows non-static strings to be used for version / help texts.

Also seen that after your comment, I'll take a look at it! Thanks for the update.

**Comment by xJonathanLEI** [comment]
Superseded by https://github.com/xJonathanLEI/starkli/commit/ddfb5aaa3a0642cf7e2df783104fcf7e9cd2d9a2.
